File: hololive/hololive-rs/crates/alarm/service/src/tier/policies.rs
use alarm_core::constants::{
    NO_UPCOMING_INTERVAL, RECENTLY_NOTIFIED_WINDOW, TIER1_INTERVAL, TIER1_WINDOW, TIER2_INTERVAL,
    TIER2_WINDOW, TIER3_INTERVAL, TIER3_WINDOW, TIER4_INTERVAL,
};
use chrono::{DateTime, Utc};

pub(super) fn compute_next_check_at(
    nearest_start: Option<DateTime<Utc>>,
    last_notified_at: Option<DateTime<Utc>>,
) -> DateTime<Utc> {
    let now = Utc::now();

    let Some(start) = nearest_start else {
        // 예정 없음: 최근 알림 발송 채널은 고빈도 폴링 유지
        if let Some(notified) = last_notified_at {
            let elapsed = now - notified;
            let window = chrono::Duration::from_std(RECENTLY_NOTIFIED_WINDOW).unwrap();
            if elapsed <= window {
                return now + chrono::Duration::from_std(TIER2_INTERVAL).unwrap();
            }
        }
        return now + chrono::Duration::from_std(NO_UPCOMING_INTERVAL).unwrap();
    };

    let time_to_start = start - now;

    // 이미 지남 또는 현재 → Tier1
    if time_to_start <= chrono::Duration::zero() {
        return now + chrono::Duration::from_std(TIER1_INTERVAL).unwrap();
    }

    let tier1 = chrono::Duration::from_std(TIER1_WINDOW).unwrap();
    let tier2 = chrono::Duration::from_std(TIER2_WINDOW).unwrap();
    let tier3 = chrono::Duration::from_std(TIER3_WINDOW).unwrap();

    if time_to_start <= tier1 {
        now + chrono::Duration::from_std(TIER1_INTERVAL).unwrap()
    } else if time_to_start <= tier2 {
        now + chrono::Duration::from_std(TIER2_INTERVAL).unwrap()
    } else if time_to_start <= tier3 {
        now + chrono::Duration::from_std(TIER3_INTERVAL).unwrap()
    } else {
        now + chrono::Duration::from_std(TIER4_INTERVAL).unwrap()
    }
}

File: hololive/hololive-shared/pkg/service/alarm/tier/policies.go
// Copyright (c) 2025 Kapu
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

package tier

import (
	"time"

	"github.com/kapu/hololive-shared/pkg/constants"
)

// ComputeNextCheckAt: 방송 시작까지 남은 시간 기반으로 다음 체크 시각을 계산한다.
// nearestStart가 nil이면 예정 없음 로직을 적용한다.
func ComputeNextCheckAt(nearestStart *time.Time, lastNotifiedAt *time.Time) time.Time {
	now := time.Now()

	if nearestStart == nil {
		// 예정 없음: 최근 알림 발송 채널은 고빈도 폴링 유지
		if lastNotifiedAt != nil {
			elapsed := now.Sub(*lastNotifiedAt)
			if elapsed <= constants.RecentlyNotifiedWindow {
				return now.Add(constants.Tier2Interval)
			}
		}
		return now.Add(constants.NoUpcomingInterval)
	}

	timeToStart := nearestStart.Sub(now)

	// 이미 지남 또는 현재 -> Tier1
	if timeToStart <= 0 {
		return now.Add(constants.Tier1Interval)
	}

	switch {
	case timeToStart <= constants.Tier1Window:
		return now.Add(constants.Tier1Interval)
	case timeToStart <= constants.Tier2Window:
		return now.Add(constants.Tier2Interval)
	case timeToStart <= constants.Tier3Window:
		return now.Add(constants.Tier3Interval)
	default:
		return now.Add(constants.Tier4Interval)
	}
}
